File: internal/sandbox/runtimes/golang/entrypoint.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
)

// Sandbox entrypoint for Go execution.
//
// Protocol:
//
//	stdin  → JSON { code, input, run_input, context, config }
//	stdout → JSON output (from output() call)
//	stderr → logs / errors
//
// User code runs inside main(). Available:
//
//	input (any), run_input (any), context (map[string]any), config (map[string]string)
//	output(val any) — call to produce result
//	fmt.Println → stderr (stdout reserved for JSON output)

type Payload struct {
	Code     string            `json:"code"`
	Input    json.RawMessage   `json:"input"`
	RunInput json.RawMessage   `json:"run_input"`
	Context  json.RawMessage   `json:"context"`
	Config   map[string]string `json:"config"`
}

func main() {
	var p Payload
	if err := json.NewDecoder(os.Stdin).Decode(&p); err != nil {
		fmt.Fprintf(os.Stderr, "entrypoint: invalid JSON payload: %v\n", err)
		os.Exit(1)
	}

	inputJSON := string(p.Input)
	if inputJSON == "" {
		inputJSON = "null"
	}
	runInputJSON := string(p.RunInput)
	if runInputJSON == "" {
		runInputJSON = "null"
	}
	contextJSON := string(p.Context)
	if contextJSON == "" {
		contextJSON = "{}"
	}
	configBytes, _ := json.Marshal(p.Config)
	if p.Config == nil {
		configBytes = []byte("{}")
	}

	tmpDir, err := os.MkdirTemp("", "sandbox-*")
	if err != nil {
		fmt.Fprintf(os.Stderr, "entrypoint: %v\n", err)
		os.Exit(1)
	}
	defer os.RemoveAll(tmpDir)

	// Write data files so user code reads them (avoids string escaping issues)
	inputPath := filepath.Join(tmpDir, "input.json")
	runInputPath := filepath.Join(tmpDir, "run_input.json")
	contextPath := filepath.Join(tmpDir, "context.json")
	configPath := filepath.Join(tmpDir, "config.json")

	_ = os.WriteFile(inputPath, []byte(inputJSON), 0644)
	_ = os.WriteFile(runInputPath, []byte(runInputJSON), 0644)
	_ = os.WriteFile(contextPath, []byte(contextJSON), 0644)
	_ = os.WriteFile(configPath, configBytes, 0644)

	// Write go.mod for the user program
	goMod := "module sandbox\ngo 1.22\n"
	_ = os.WriteFile(filepath.Join(tmpDir, "go.mod"), []byte(goMod), 0644)

	// Generate user source
	src := `package main

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"sort"
	"strconv"
	"strings"
	"time"
)

var _realStdout = os.Stdout

func init() {
	os.Stdout = os.Stderr
}

func output(val any) {
	b, err := json.Marshal(val)
	if err != nil {
		fmt.Fprintf(os.Stderr, "output serialization error: %v\n", err)
		_realStdout.Write([]byte("null"))
	} else {
		_realStdout.Write(b)
	}
	os.Exit(0)
}

// Suppress unused import errors
var (
	_ = math.Pi
	_ = sort.Ints
	_ = strconv.Itoa
	_ = strings.Contains
	_ = time.Now
)

func main() {
	var input any
	inputBytes, _ := os.ReadFile(` + quote(inputPath) + `)
	json.Unmarshal(inputBytes, &input)

	var run_input any
	runInputBytes, _ := os.ReadFile(` + quote(runInputPath) + `)
	json.Unmarshal(runInputBytes, &run_input)
	_ = run_input

	context := map[string]any{}
	contextBytes, _ := os.ReadFile(` + quote(contextPath) + `)
	json.Unmarshal(contextBytes, &context)

	config := map[string]string{}
	configBytes, _ := os.ReadFile(` + quote(configPath) + `)
	json.Unmarshal(configBytes, &config)

	_ = input
	_ = context
	_ = config

` + p.Code + `

	_realStdout.Write([]byte("null"))
}
`

	srcPath := filepath.Join(tmpDir, "main.go")
	if err := os.WriteFile(srcPath, []byte(src), 0644); err != nil {
		fmt.Fprintf(os.Stderr, "entrypoint: write source: %v\n", err)
		os.Exit(1)
	}

	cmd := exec.Command("go", "run", ".")
	cmd.Dir = tmpDir
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	if err := cmd.Run(); err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			os.Exit(exitErr.ExitCode())
		}
		os.Exit(1)
	}
}

// quote returns a Go double-quoted string literal.
func quote(s string) string {
	b, _ := json.Marshal(s)
	return string(b)
}

File: internal/sandbox/runtimes/javascript/entrypoint.js
#!/usr/bin/env node
/**
 * Sandbox entrypoint for JavaScript/TypeScript execution.
 *
 * Protocol:
 *   stdin  → JSON { code, input, context, params }
 *   stdout → JSON output (result of user code)
 *   stderr → logs / errors
 *
 * User code runs as ESM (.mjs) via dynamic `await import()` in BOTH run and
 * debug modes. Globals available in user code:
 *   input, context, params, output, console
 * Module loading: `const x = await import('pkg')`. CommonJS `require` is
 * intentionally not provided — ESM scope makes it impractical to expose,
 * and unifying on `await import` keeps run/debug behavior identical.
 */
'use strict';

const fs = require('fs');
const { spawn } = require('child_process');

process.stdin.setEncoding('utf8');

let raw = '';
process.stdin.on('data', (chunk) => { raw += chunk; });

process.stdin.on('end', async () => {
  let payload;
  try {
    payload = JSON.parse(raw);
  } catch (err) {
    process.stderr.write(`entrypoint: invalid JSON payload: ${err.message}\n`);
    process.exit(1);
  }

  const { code, input, context, params } = payload;

  // Setup script: sets globals via globalThis, loaded via -r flag (CJS).
  // Used by both run and debug modes.
  // NOTE (debug): do NOT override console — real console.log must fire so CDP
  // Runtime.consoleAPICalled events reach the debug UI console panel.
  // NOTE (run): console.log is redirected to stderr inside the runner so
  // stdout stays clean for the final JSON result.
  const setupScript = `'use strict';
globalThis.input = ${JSON.stringify(input ?? null)};
globalThis.context = ${JSON.stringify(context ?? {})};
globalThis.params = ${JSON.stringify(params ?? {})};
globalThis._outputValue = undefined;
globalThis._outputSet = false;
globalThis.output = (val) => { globalThis._outputValue = val; globalThis._outputSet = true; };
`;

  // User script: pure user code, line 1 = user line 1 (preserves breakpoints).
  const userScript = code;

  if (process.env.SANDBOX_DEBUG) {
    // Debug mode: spawn child with --inspect-brk, real console for CDP events.
    const runnerScript = `'use strict';
process.on('uncaughtException', (err) => {
  console.error('uncaught exception:', err.message);
  if (err.stack) console.error(err.stack);
  setTimeout(() => process.exit(1), 50);
});
process.on('unhandledRejection', (reason) => {
  const msg = reason instanceof Error ? reason.message : String(reason);
  const stack = reason instanceof Error ? reason.stack : undefined;
  console.error('unhandled rejection:', msg);
  if (stack) console.error(stack);
  setTimeout(() => process.exit(1), 50);
});

(async () => {
  try {
    await import('file:///sandbox/user_script.mjs');
    await new Promise(r => setTimeout(r, 50));
    const _result = globalThis._outputSet ? globalThis._outputValue : null;
    process.stdout.write(JSON.stringify(_result));
    // Tagged emission so CDP/Go can separate final result from console output
    console.log('__SANDBOX_RESULT:' + JSON.stringify(_result, null, 2));
    process.exitCode = 0;
  } catch (err) {
    console.error('runtime error:', err.message);
    if (err.stack) console.error(err.stack);
    process.exitCode = 1;
  }
})();
`;

    fs.writeFileSync('/sandbox/setup.js', setupScript);
    fs.writeFileSync('/sandbox/user_script.mjs', userScript);
    fs.writeFileSync('/sandbox/runner.js', runnerScript);

    process.stderr.write('node: starting with --inspect-brk on port 9229...\n');

    const child = spawn('node', [
      '--inspect-brk=0.0.0.0:9229',
      '-r', '/sandbox/setup.js',
      '/sandbox/runner.js',
    ], {
      stdio: ['ignore', 'pipe', 'pipe'],
    });

    child.stdout.on('data', (d) => process.stdout.write(d));
    child.stderr.on('data', (d) => process.stderr.write(d));
    child.on('exit', (exitCode) => process.exit(exitCode ?? 1));
    return;
  }

  // Run mode: redirect console to stderr (stdout reserved for JSON result),
  // load globals, then dynamic-import user code as ESM.
  const origLog = console.log;
  console.log   = (...args) => process.stderr.write(args.map(String).join(' ') + '\n');
  console.error = (...args) => process.stderr.write(args.map(String).join(' ') + '\n');
  console.warn  = (...args) => process.stderr.write(args.map(String).join(' ') + '\n');
  console.info  = (...args) => process.stderr.write(args.map(String).join(' ') + '\n');
  void origLog;

  globalThis.input    = input ?? null;
  globalThis.context  = context ?? {};
  globalThis.params   = params ?? {};
  globalThis._outputValue = undefined;
  globalThis._outputSet   = false;
  globalThis.output = (val) => { globalThis._outputValue = val; globalThis._outputSet = true; };

  fs.writeFileSync('/sandbox/user_script.mjs', userScript);

  // 30s safety net (matches prior vm timeout)
  const watchdog = setTimeout(() => {
    process.stderr.write('runtime error: script exceeded 30s timeout\n');
    process.exit(1);
  }, 30000);
  watchdog.unref();

  try {
    await import('file:///sandbox/user_script.mjs');
    const result = globalThis._outputSet ? globalThis._outputValue : null;
    process.stdout.write(JSON.stringify(result));
    process.exit(0);
  } catch (err) {
    process.stderr.write(`runtime error: ${err.message}\n`);
    if (err.cause) {
      const c = err.cause;
      const causeMsg = c instanceof Error ? `${c.name}: ${c.message}${c.code ? ` (${c.code})` : ''}` : String(c);
      process.stderr.write(`caused by: ${causeMsg}\n`);
      if (c.stack) process.stderr.write(c.stack + '\n');
    }
    if (err.stack) {
      process.stderr.write(err.stack + '\n');
    }
    process.exit(1);
  }
});
